pkg/plugins/policies/meshpassthrough/plugin/xds/order.go

```diff
@@ -11,109 +13,183 @@ import (
+		matchType, isWildcardDomain := getMatchType(match, protocol)
+		matcher := Matcher{
+			Protocol:  protocol,
+			Port:      port,
+			MatchType: matchType,
+		}
 		if _, found := portProtocols[port]; !found {
 			portProtocols[port] = map[core_mesh.Protocol]bool{protocol: true}
 		} else {
 			portProtocols[port][protocol] = true
 		}
-		matcher := Matcher{
-			Protocol: protocol,
-			Port:     uint32(port),
-		}
-		route := Route{
-			Value:     match.Value,
-			MatchType: matchType,
-		}
-		if _, found := matchers[matcher]; found {
-			matchers[matcher][route] = true
-		} else {
-			matchers[matcher] = map[Route]bool{
-				route: true,
+		switch protocol {
+		case core_mesh.ProtocolHTTP, core_mesh.ProtocolHTTP2, core_mesh.ProtocolGRPC:
+			// when there are domains we want to create VirtualHosts with Domain match
+			if matchType == Domain {
+				if isWildcardDomain {
+					matchType = WildcardDomain
+				}
+				route := Route{
+					Value:     match.Value,
+					MatchType: matchType,
+				}
+				if _, found := matcherWithRoutes[matcher]; found {
+					matcherWithRoutes[matcher][route] = true
+				} else {
+					matcherWithRoutes[matcher] = map[Route]bool{
+						route: true,
+					}
+				}
+			} else {
+				matcher.Value = match.Value
+				// there should be no existing matcher if there is ip/cidr
+				matcherWithRoutes[matcher] = map[Route]bool{}
 			}
+		default:
+			matcher.Value = match.Value
+			matcherWithRoutes[matcher] = map[Route]bool{}
 		}
 	}
-	filterChainMatchers := []FilterChainMatcher{}
-	for matcher, routesMap := range matchers {
-		routes := []Route{}
-		for route := range routesMap {
-			routes = append(routes, route)
-		}
-		orderRoutes(routes)
-		filterChainMatcher := FilterChainMatcher{
-			Protocol: matcher.Protocol,
-			Port:     matcher.Port,
-			Routes:   routes,
-		}
-		filterChainMatchers = append(filterChainMatchers, filterChainMatcher)
-	}
+	// we cannot differentiate between HTTP, HTTP/2, and gRPC on the same port.
 	if err := validatePortAndProtocol(portProtocols); err != nil {
 		return nil, err
 	}
-	orderValues(filterChainMatchers)
+	// Envoy first checks the port when performing matching. If there is a matcher for a specific port
+	// and one rule to match all ports alongside another for a specific port,
+	// it might select the matcher for the specific port but fail to find a corresponding filter chain.
+	// To avoid this issue, we also generate specific port matchers for rules intended to match all ports.
+	matcherWithRoutesAndAdditionalPorts := map[Matcher]map[Route]bool{}
+	for matcher, routes := range matcherWithRoutes {
+		if _, found := matcherWithRoutesAndAdditionalPorts[matcher]; found {
+			for route := range routes {
+				matcherWithRoutesAndAdditionalPorts[matcher][Route{
+					Value:     route.Value,
+					MatchType: route.MatchType,
+				}] = true
+			}
+		} else {
+			matcherWithRoutesAndAdditionalPorts[matcher] = maps.Clone(routes)
+		}
```

Comment:
Do we have the `found` case? 
Since `matcherWithRoutesAndAdditionalPorts` and `matcherWithRoutes` are the same structure `map[Matcher]map[Route]bool{}`, a map only has one unique key right?

Please correct me if I'm wrong